pkg/agent/billing.go

```diff
@@ -0,0 +1,181 @@
+package agent
+
+import (
+	"context"
+	"fmt"
+	"net/http"
+	"time"
+
+	"github.com/google/uuid"
+
+	"k8s.io/apimachinery/pkg/types"
+
+	vmapi "github.com/neondatabase/neonvm/apis/neonvm/v1"
+
+	"github.com/neondatabase/autoscaling/pkg/billing"
+	"github.com/neondatabase/autoscaling/pkg/util"
+)
+
+type BillingConfig struct {
+	URL                 string `json:"url"`
+	CollectEverySeconds uint   `json:"collectEverySeconds"`
+	PushEverySeconds    uint   `json:"pushEverySeconds"`
+	PushTimeoutSeconds  uint   `json:"pushTimeoutSeconds"`
+}
+
+type billingMetricsState struct {
+	historical map[billingMetricsKey]vmMetricsHistory
+	present    map[billingMetricsKey]vmMetrics
+	lastTime   *time.Time
+}
+
+type billingMetricsKey struct {
+	uid        types.UID
+	endpointID string
+}
+
+type vmMetricsHistory struct {
+	slices []metricsTimeSlice
+}
+
+type metricsTimeSlice struct {
+	metrics   vmMetrics
+	startTime time.Time
+	endTime   time.Time
+}
+
+type vmMetrics struct {
+	cpu uint16
+}
+
+const (
+	MetricNameCPU string = "cpu"
+	EndpointLabel string = "neon/endpoint-id"
+)
+
+func RunBillingMetricsColllector(backgroundCtx context.Context, conf *BillingConfig, store *util.WatchStore[vmapi.VirtualMachine]) error {
+	client := billing.NewClient(conf.URL, http.DefaultClient)
+
+	collectTicker := time.NewTicker(time.Second * time.Duration(conf.CollectEverySeconds))
+	defer collectTicker.Stop()
+	pushTicker := time.NewTicker(time.Second * time.Duration(conf.PushEverySeconds))
+	defer pushTicker.Stop()
+
+	state := billingMetricsState{
+		historical: make(map[billingMetricsKey]vmMetricsHistory),
+		present:    make(map[billingMetricsKey]vmMetrics),
+		lastTime:   nil,
+	}
+
+	state.collect(conf, store)
+
+	for {
+		select {
+		case <-collectTicker.C:
+			state.collect(conf, store)
+		case <-pushTicker.C:
+			batch := state.drainIntoBatch(&client)
+			if err := pushBillingEvents(conf, batch); err != nil {
+				return fmt.Errorf("Error pushing billing events: %w", err)
```

Comment:
does this mean that if we fail to push once we stop collecting consumption data? 

While I don't expect the console to be buggy, this seems aggressive.

Reply:
Yeah, that's what it'll currently do ­— originally wasn't sure about how best to resolve that and was planning on asking you. Although now I think the best way is probably just saving the batch for later if it fails, then re-sending with the merged batch on the next try? new code would look like:

```go
batch := client.NewBatch()
for {
    // -- skipped --
    case <-pushTicker.C:
        state.drainAppendToBatch(&batch)
        if err := pushBillingEvents(conf, batch); err != nil {
            klog.Errorf(...)
            continue
        }
        // reset batch, because sending was successful
        batch = client.NewBatch()
    // -- skipped --
}
```

Reply:
handled in latest commit, as previously discussed.